File: services/user-service/src/main/java/com/visor/school/userservice/bootstrap/KeycloakInitializerRunner.java
package com.visor.school.userservice.bootstrap;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.ApplicationArguments;
import org.springframework.boot.ApplicationRunner;
import org.springframework.stereotype.Component;

import com.visor.school.keycloak.InitializerService;
import com.visor.school.keycloak.config.InitializerProperties;
import com.visor.school.keycloak.model.InitializationOutcome;
import com.visor.school.keycloak.model.KeycloakBlueprint;

@Component
public class KeycloakInitializerRunner implements ApplicationRunner {

    private final Logger log = LoggerFactory.getLogger(getClass());
    private final InitializerService initializerService;
    private final InitializerProperties properties;

    public KeycloakInitializerRunner(InitializerService initializerService, InitializerProperties properties) {
        this.initializerService = initializerService;
        this.properties = properties;
    }

    @Override
    public void run(ApplicationArguments args) throws Exception {
        if (!properties.isEnabled()) {
            log.info("Keycloak initializer disabled via configuration; skipping startup provisioning");
            return;
        }

        KeycloakBlueprint blueprint = properties.toBlueprint();
        InitializerProperties.Retry retry = properties.getRetry();
        int attempt = 1;
        long backoff = retry.getInitialBackoffMillis();

        while (attempt <= retry.getMaxAttempts()) {
            try {
                InitializationOutcome outcome = initializerService.initialize(blueprint);
                log.info("Keycloak initializer outcome: {}", outcome);
                return;
            } catch (Exception ex) {
                if (attempt == retry.getMaxAttempts()) {
                    if (retry.isFailOnError()) {
                        log.error("Keycloak initialization failed after {} attempts", attempt, ex);
                        throw ex;
                    } else {
                        log.warn(
                            "Keycloak initialization failed after {} attempts. Application will start without Keycloak initialization. Error: {}",
                            attempt,
                            ex.getMessage()
                        );
                        return;
                    }
                }

                log.warn(
                    "Keycloak initialization attempt {} failed: {}. Retrying in {} ms",
                    attempt,
                    ex.getMessage(),
                    backoff
                );

                try {
                    Thread.sleep(Math.max(100L, backoff));
                } catch (InterruptedException interrupted) {
                    Thread.currentThread().interrupt();
                    if (retry.isFailOnError()) {
                        throw ex;
                    } else {
                        log.warn("Keycloak initialization interrupted. Application will start without Keycloak initialization.");
                        return;
                    }
                }

                backoff = (long) (backoff * retry.getMultiplier());
                attempt++;
            }
        }
    }
}

File: services/user-service/src/main/kotlin/com/visor/school/userservice/bootstrap/KeycloakInitializerRunner.kt
package com.visor.school.userservice.bootstrap

import com.visor.school.keycloak.InitializerService
import com.visor.school.keycloak.config.InitializerProperties
import org.slf4j.LoggerFactory
import org.springframework.boot.ApplicationArguments
import org.springframework.boot.ApplicationRunner
import org.springframework.stereotype.Component
import kotlin.math.max

@Component
class KeycloakInitializerRunner(
    private val initializerService: InitializerService,
    private val properties: InitializerProperties
) : ApplicationRunner {

    private val log = LoggerFactory.getLogger(javaClass)

    override fun run(args: ApplicationArguments) {
        if (!properties.enabled) {
            log.info("Keycloak initializer disabled via configuration; skipping startup provisioning")
            return
        }

        val blueprint = properties.toBlueprint()
        val retry = properties.retry
        var attempt = 1
        var backoff = retry.initialBackoffMillis

        while (attempt <= retry.maxAttempts) {
            try {
                val outcome = initializerService.initialize(blueprint)
                log.info("Keycloak initializer outcome: {}", outcome)
                return
            } catch (ex: Exception) {
                if (attempt == retry.maxAttempts) {
                    if (retry.failOnError) {
                        log.error("Keycloak initialization failed after {} attempts", attempt, ex)
                        throw ex
                    } else {
                        log.warn(
                            "Keycloak initialization failed after {} attempts. Application will start without Keycloak initialization. Error: {}",
                            attempt,
                            ex.message
                        )
                        return
                    }
                }

                log.warn(
                    "Keycloak initialization attempt {} failed: {}. Retrying in {} ms",
                    attempt,
                    ex.message,
                    backoff
                )

                try {
                    Thread.sleep(max(100L, backoff))
                } catch (interrupted: InterruptedException) {
                    Thread.currentThread().interrupt()
                    if (retry.failOnError) {
                        throw ex
                    } else {
                        log.warn("Keycloak initialization interrupted. Application will start without Keycloak initialization.")
                        return
                    }
                }

                backoff = (backoff * retry.multiplier).toLong()
                attempt++
            }
        }
    }
}
